Initial `Target` C API

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Fixes #14241 and #14376

The following commits enable a C FFI for creation of a Rust `Target`.


### Details and comments
As we work into expanding Qiskit into working in C, it is important to bring parts of the transpiler into this domain. One such part is the `Target`, which contains a lot of information about the hardware a circuit will be executed in. The following commits expose a C FFI that allows us to create and interact with a Target directly from C.

In order to achieve this we have 3 types of pointers being exposed to C:

* `QkTarget` - which points to a `Target` in Rust.
* `QkPropsMap` - which represents a mapping between `Qargs` : `InstructionProperties`, or, in C terms, this would be a mapping of `uint32_t[]` : `QkInstructionProps`.

The C FFI should allow us to:
- Fully build a Target based on standard gates ``QkGate``, a list of parameters `double[]`, and an optional `QkPropsMap` containing the instruction properties.
- Update any of the `QkTarget` instruction's properties after being added.

raynelfss committed:
Initial: Enable a C API to build and modify a Target

raynelfss committed:
Add: `add_instruction` function

raynelfss committed:
Add: More methods to `Target` C API
- Add provisional `PropertyMap` to represent the mapping of `qargs : InstructionProperties` before adding it to the `Target`.
- `InstructionProperties`:
    - free.
- `Target`:
    - `update_instruction_prop`
    - `operation_names`
    - `physical_qubits`
    - `non_global_operation_names`.
    - `operation_names_for_qargs`.
    - `qargs_for_operation_names`.
    - `qargs`.
    - `instruction_supported`.
    - `contains_key`.

raynelfss committed:
Add: C unit test
- Add getter methods for `InstructionProperties`.
- Add `length` method for `PropertyMap`.
- Add name maping for `PropertyMap`.

raynelfss committed:
Fix: Add `IndexMap` and `Ahash` dependencies to qiskit_cext.
- Add `get` method to `Target`.
- Add missing methods to `PropertyMap`.
- Fix methods returning arrays to C by adding a a forget statement to prevent pointer from being destroyed.
- Add getter methods to the `C` api of target.
- Add helper function to parse qargs.

raynelfss committed:
Test: Add tests to update properties and get non global names.

raynelfss committed:
Add `operation_names_for_qargs` test case.
- Move `print_qargs` helper to top of module.

raynelfss committed:
Add: `qargs_for_operation_names` test case
- Add helper method to compare arrays.
- Fix print method for qargs.

raynelfss committed:
Docs: Add first wave of docstrings.

raynelfss committed:
Fix: Incorrect typing in `test_target.c`

raynelfss committed:
Fix: Add missing docstrings.
- Use correct type of pointers for outputs,

raynelfss committed:
Lint: Fix c-format

raynelfss committed:
Fix: Remove option from `qarg_gate_map`.
- Use IndexSet to preserve insertion order to avoid sorting in the C API.

raynelfss committed:
Test: Add `qargs` test.

raynelfss committed:
FIx: Add missing docstrings for remainign tests

raynelfss committed:
Fix: Type errors in MacOS

raynelfss committed:
Fix: Undo extra changes in the target.
- Sort the string arrays obtained from the target in C during tests.
- Remove `Copy` from `InstructionProperties`.
- Add helper method to sort string array outputs, coming from the Target.

raynelfss committed:
Add: more complete constructor for `Target`.
- Add getters for `description`, `dt` `min_length` `granularity`, `acquire_alignment`, and `pulse_alignment`.
- Add Target construction test.

mtreinish commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1296 @@
+    granularity: *const u32,
+    min_length: *const usize,
+    pulse_alignment: *const u32,
+    acquire_alignment: *const u32,
```
This is a really weird pattern for these arguments. I get that you're trying to have `Option<T>` here by using the null pointer, but treating these all as pointers is not a good way to do this. Instead I would just have methods to set these. I don't think `num_qubits` being `None` is something we need in C, it's a weird edge case that doesn't exist in real targets. So I would make this signature `qk_target_new(num_qubits: u32)`. Then add a bunch of new functions for `qk_target_set_dt(target: *mut Target, dt: f64)`, etc.

mtreinish reviewed (commented):
I know this isn't ready for review yet, and I haven't done a full review. But I opened this up and got triggered on the first c function so left a comment there.

raynelfss committed:
Fix: Clean up pointers in `qk_target_new`
- Add setters for `QkTarget` attributes.

raynelfss committed:
Fix: Add release note

raynelfss committed:
Fix: Typos in docstring

raynelfss committed:
Lint: Fix lint for C

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1418 @@
+/// # Example
+///     
+///     QkTarget *target = qk_target_new("New Target", 5);
+///     char *description = qk_target_description(target);
```
The string would need to be freed after usage, right? It would be good to include this here, maybe print and then free the `char*`?

Cryoris replied:
Edit: I believe `description` needs to be freed Rust side (where the string was allocated) via
```
qk_str_free(description);
```

raynelfss replied:
Since usage of strings was removed in 46cc693, this is no longer necessary.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1418 @@
+///
+/// # Example
+///     const dt = 10e-9;
+///     QkTarget *target = qk_target_new("New Target", 5, &dt, NULL, NULL, NULL, NULL);
```
This example looks out of date.

There are some other cases of this below as well.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1418 @@
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_dt(target: *const Target) -> *mut f64 {
```
Hmmm, it's not amazing that we have to return a pointer here to handle the `dt=None` case, since then we'd also have to free the `f64` later on again. Another solution would be to take a `dt: *mut f64` pointer as input and only set if if `dt` is set.

raynelfss replied:
This is also out of date, the plan is to return an f64 value. I can return 0 here for the case of `None`.

raynelfss replied:
Reminiscing here, maybe it should return `NaN` as it would make more sense rather than implying 0.

raynelfss replied:
Fixed in 799a71f

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1418 @@
+use std::ptr::null_mut;
+
+/// @ingroup QkTarget
+/// Construct a new Target with the given number of qubits.
```
It would be nice to consistently have 
```
``Target``
```
(or double quotes for other args/code words 🙂)

raynelfss replied:
Fixed in 6ac8072

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1418 @@
+    // SAFETY: Per documentation, name points to a null-terminated string of characters.
+    let description: &CStr = unsafe { CStr::from_ptr(description) };
+    target.description = Some(description.to_str().unwrap().to_string());
+    ExitCode::Success
```
Since this always either returns `Success` or fails I don't think we need an exit code here 🤔  but I'm also fine keeping it if you really want to, but then it should add it to the docs 🙂 

This comment also applies to some other functions below.

raynelfss replied:
I'm only putting it here because of the syntax of `ExitCode`. IIRC, `ExitCode` should be used when mutating things. If that's not the case, I would be happy to remove it.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1418 @@
+    {
+        ExitCode::Success
+    } else {
+        ExitCode::CInputError
```
We can also add a custom error code, `CInputError` is very generically saying that the input from C was faulty. Maybe we can have a transpiler-specific set of error codes with target errors being part of them?

raynelfss replied:
Fixed in 46cc693

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1418 @@
+pub unsafe extern "C" fn qk_target_non_global_operation_names(
+    target: *mut Target,
+    strict_direction: bool,
+) -> *mut *mut c_char {
```
This looks very strange, how do we know the length of the outer vector? Would it make more sense to have a dedicated struct that users can query from? 

Another approach could also be to not use `*char` to describe the operations, since in C we only have the standard gates anyways. Maybe we can also discuss this offline.

raynelfss replied:
A struct would make more sense here since we can store the length of the output. I will make sure to add that.

raynelfss replied:
Fixed in a9ad527

Cryoris commented on releasenotes/notes/c-api-target-df63634954652fac.yaml:
```diff
@@ -0,0 +1,5 @@
+features_transpiler:
+  - |
+    Support for creation, manipulation, and interaction of :class:`~.Target` have
+    been added to the Qiskit C API.
```
Once we're happy with the API we should add an example here 🙂

Cryoris replied:
We are happy now I think 🙂

Cryoris reviewed (commented):
Really nice to see this! Here's a super high-level set of comments regarding the syntax mostly, I'll have a closer look later 🙂 

Also: we also need to add the `QkTarget` group (and the other objects) to the cdocs

raynelfss committed:
Merge branch 'main' into target-c-api

raynelfss committed:
Fix: Use structs for collections, specify length.
- Introduce `QkTargetQargs`, `QkTargetNameList`, and `QkTargetQargsList` to expose collections of operation names and qargs to C exposing the length of the collection.

raynelfss committed:
Fix: Use ``Target`` for docstrings.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-api

raynelfss committed:
Fix: Use the `qiskit_transpiler` crate.

raynelfss committed:
Fix: Add "qiskit-transpiler" to cbindgen.toml
- Remove incorrect usage of pointers for `qk_target_dt`.

raynelfss committed:
Merge branch 'main' into target-c-api

raynelfss commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1459 @@
+/// pointer is not valid or not null ending, the behavior will be undefined.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_new(description: *mut c_char, num_qubits: usize) -> *mut Target {
```
Me and @Cryoris were discussing about whether a description is still needed for a `Target` in the C API. Is there any use for it currently?

raynelfss replied:
After internal discussion, description is no longer part of the initial C API, removed in e5911c3

raynelfss committed:
Fix: Incorrect import of `PhysicalQubit` in target.rs

raynelfss committed:
Fix: Remove usage of `string` to query instructions, use `StandardGate` instead.
- Introduce `Target` specific exitcodes starting at 300.
- Change methods to use `StandardGate` instead of String.

raynelfss committed:
Fix: Remove `description` from the C API.

raynelfss committed:
FIx: Remove `InstructionProperties` from the C API
- Introduce `QkInstProperties` as a C representable variant of `InstructionProperties`.
- Move `InstructionProperties` instantiation to function calls. Now you can directly send duration or error as function arguments.
- Make `PropsMap` be either initialized or `NULL`.

raynelfss committed:
Fix: Incorrect documentation usage for `qk_target_new()`

raynelfss committed:
Fix: Use NAN in one of the examples.

raynelfss committed:
Docs: Add documentation for the `Target` C API

raynelfss committed:
Lint: Qiskit header format in docs

raynelfss committed:
Fix: Use `NaN` when `dt` is not set.

raynelfss committed:
Merge branch 'main' into target-c-api

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,1278 @@
+/// @return The dt value of this ``Target`` or `NaN` if not assigned.
+///
+/// # Example
+///     const dt = 10e-9;
```
It looks like this needs to be removed 🙂 
```suggestion
///
```

raynelfss replied:
Removed in 395a194

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-api

raynelfss committed:
Refactor: Remove all querrying methods from Target, PropsMap.
- Remove all helper structs.
- All removed changes will be spun off to a different PR.

raynelfss committed:
Fix: Implement `From<TargetError>` for `ErrorCode`.

raynelfss committed:
Add: Separate method to add gates with fixed params.
- Since parameters are not fully supported in C as of yet, we have to separate the methods to make a clear division between how to add a gate with/without parameters, and making additions with fixed parameters fully explicit in the Target.

raynelfss committed:
Docs: Address new changes in documentation

raynelfss committed:
Fix: Add a way of copying a Target from C.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+}
+
+/// @ingroup QkTarget
+/// Returns the `min_length` value of this ``Target``.
```
There are also some other places where this misses the second quote 🙂 
```suggestion
/// Returns the ``min_length`` value of this ``Target``.
```

raynelfss replied:
Addressed in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// @return The min_length value of this ``Target``.
+///
+/// # Example
+///     QkTarget *target = qk_target_new(5);
```
I don't know if this needs a linebreak in between the header and the code -- could you verify in the docs build if this is rendered correctly? 🙂

raynelfss replied:
Fixed in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+///
+/// @param target A pointer to the ``Target``.
+///
+/// @return The min_length value of this ``Target``.
```
Some more doc nits 😉 This also applies to some other return statements below, e.g. ``pulse_alignment``
```suggestion
/// @return The ``min_length`` value of this ``Target``.
```

raynelfss replied:
Addressed in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// Sets the dt value of this ``Target``.
+///
+/// @param target A pointer to the ``Target``.
+/// @param dt The dt value for the system time resolution of input.
```
The setters are missing the return docs (they return an `ExitCode`)

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+///
+/// # Example
+///     QkTarget *target = qk_target_new(5);
+///     double dt = qk_target_set_dt(target, 10e-9);
```
This should return an exit code

raynelfss replied:
Added in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// non-null pointer to a ``QkPropsMap`` object.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_property_map_length(property_map: *const PropertyMap) -> usize {
```
Could we call this `len` instead of `length` for consistency with `qk_obs_len`?

raynelfss replied:
Renamed in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+    property_map: *mut PropertyMap,
+    qargs: *mut u32,
+    num_qubits: u32,
+    duration: f64,
```
I wonder if we want to do some input validation here -- e.g. the duration can't be below the min length of the target, or the error must be in [0, 1]. Do Python or Rust do input validation?

raynelfss replied:
They don't do any input validation as far as I'm aware. If you believe this might be something we wanna add, you should open an issue. It might be helpful to add this to the Target.

Cryoris replied:
Sounds good, let's keep as is then

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+}
+
+/// @ingroup QkTarget
+/// Adds a non-parameteric StandardGate to the ``Target``.
```
The same comment also in some other locations below: standard gates are exposed as the enum ``QkGate``
```suggestion
/// Adds a non-parameteric ``QkGate`` to the ``Target``.
```

raynelfss replied:
Renamed in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// Adds a non-parameteric StandardGate to the ``Target``.
+///
+/// This function will fail if a parametric gate is added.
+///
```
I think this can be removed given the `@note` below
```suggestion
```

raynelfss replied:
Removed in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+///
+/// @note Adding parametric gates with non-fixed parameters is currently
+/// not supported. See ``qk_target_add_instruction_fixed_params()`` for
+/// adding gates with fixed parameters.
```
Does this render correctly or do we need an indentation for the following lines? Same question for `@param` below, and in the `qk_target_add_instruction_fixed_params` function, too

raynelfss replied:
![image](https://github.com/user-attachments/assets/ce8f7c6b-852f-48cc-b180-036bc2fe75d4)
I modified the ``QkGate`` instance here, but nothing else has been modified.

Cryoris replied:
Good to know, thanks for checking!

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// @param target A pointer to the ``Target``.
+/// @param instruction The StandardGate to be added to the ``Target``.
+/// @param property_map The mapping of qargs and InstructionProperties to
+/// be associated with this instruction.
```
This misses return docs

raynelfss replied:
How would you recomment showing the ``ExitCode`` in the return satement? Does it matter if I describe the possible error codes returned by the function?

Cryoris replied:
Just that it returns some exit code is fine I'd say, users can look up the code in the docs 🙂

raynelfss replied:
Added in the lines after these, in commit 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// Behavior is undefined if ``property_map`` is not a valid, non-null pointer to a ``QkPropsMap``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_add_instruction_fixed_params(
```
What do you think about calling this ``qk_target_add_fixed_instruction`` to reduce the length a bit?

raynelfss replied:
Renamed in 395a194

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// @param duration The instruction's duration in seconds on the specific set of
+///     qubits.
+/// @param error The instruction's average error rate on the specific set of qubits.
+/// # Example
```
Return missing (I'll stop commenting that now 🙂)

raynelfss replied:
And I'll stop saying "fixed in 395a194"

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// for a given gate.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_update_instruction_prop(
```
How about abbreviating to `qk_target_update_instruction`? Or `qk_target_update_property`? C function names are already quite long 😅

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,741 @@
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_length(target: *const Target) -> usize {
```
What is a target length? 🤔

raynelfss replied:
The number of instructions it contains, would you suggest a different name that conveys this better?

Cryoris replied:
`target_num_instructions`? 😇

Cryoris commented on docs/cdoc/qk-props-map.rst:
```diff
@@ -0,0 +1,43 @@
+.. note::
+    This feature might be removed in a future release as we move to more C native
+    structure usage.
+
```
That's true for a lot of structs, I don't think we have to point the volatility out 😛 
```suggestion
```

Cryoris commented on docs/cdoc/qk-target.rst:
```diff
@@ -0,0 +1,51 @@
+
+The Target C API currently only supports additions of ``StandardGate`` instances as
+``NormalOperation`` and queries using ``StandardGate`` instead of gate names. The
+functionality will be expanded over time as we improve our Rust data model capabilities.
```
This looks out of date

Cryoris commented on docs/cdoc/qk-target.rst:
```diff
@@ -0,0 +1,51 @@
+    qk_target_add_instruction(target, QkGate_CX, property_map);
+
+    // Add a global H gate
+    qk_target_add_instruction(target, QkGate_H, NULL);
```
This global H gate has no properties? How does this work?

raynelfss replied:
It gets added as a global gate, meaning that, for this backend, the gate is compatible for any combination of qargs (of the right size of course).

Cryoris replied:
But there's no error or duration attached, either? This seems a bit strange to me, but we can keep it for now -- the more general question is whether we actually want to force all properties to be defined or not 🤔

Cryoris commented on docs/cdoc/qk-target.rst:
```diff
@@ -0,0 +1,51 @@
+of a backend. Its purpose is to provide the compiler with information that allows it
+to compile an input circuit into another that is optimized taking in consideration the
+``Target``'s specifications. This structure represents a low level interface to the main
+:class:`.Target` data structure in Rust, which represents the base class for its Python
```
I think this would link to the Python class, not the Rust class, no?

raynelfss replied:
It does here. Although that was intended at first, I would need to rewrite it.

Cryoris commented on docs/cdoc/qk-target.rst:
```diff
@@ -0,0 +1,51 @@
+.. code-block:: C
+
+    #include <qiskit.h>
+    #include "math.h"
```
Quotes are typically used for local files
```suggestion
    #include <math.h>
```

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,315 @@
+// that they have been altered from the originals.
+
+#include "common.h"
+#include "math.h"
```
```suggestion
#include <math.h>
```

Cryoris reviewed (commented):
Thanks a lot for all the work, and for the split 🙂 This overall looks quite nice, I have mainly very small comments below. 

I have one question/suggestion on the property map -- intuitively, to me, it is a bit strange that the property map does not contain the gate it defines the properties for. I was expecting a structure more like
```c
QkInstructionProperties *inst = qk_instruction_properties_new(QkGate_H);
qk_instruction_properties_add(qargs, len, error, duration);  // add data...
qk_target_add(target, inst); 
```
or alternatively, for the fixed parameters
```c
QkInstructionProperties *inst = qk_instruction_properties_new_fixed(QkGate_RZ, params);
```
What would people think about this?

raynelfss committed:
Fix: Address review comments

Co-authored-by: Julien Gacon <gaconju@gmail.com>

raynelfss committed:
Lint: Fix formatting

raynelfss committed:
Merge branch 'main' into target-c-api

raynelfss committed:
Remove: `Target::get` method.
- Will be re-added in future updates.

raynelfss committed:
Docs: Remove volatility line from `PropsMap` docs.

raynelfss commented:
> Thanks a lot for all the work, and for the split 🙂 This overall looks quite nice, I have mainly very small comments below.
> 
> I have one question/suggestion on the property map -- intuitively, to me, it is a bit strange that the property map does not contain the gate it defines the properties for. I was expecting a structure more like
> 
> ```c
> QkInstructionProperties *inst = qk_instruction_properties_new(QkGate_H);
> qk_instruction_properties_add(qargs, len, error, duration);  // add data...
> qk_target_add(target, inst); 
> ```
> 
> or alternatively, for the fixed parameters
> 
> ```c
> QkInstructionProperties *inst = qk_instruction_properties_new_fixed(QkGate_RZ, params);
> ```
> 
> What would people think about this?

This is a nice idea, as we can initialize the property map with a gate and its parameters (when needed) instead of going through the extra steps. For this, I assume, we want to expand the `PropertyMap` structure into something different (perhaps it can become `TargetEntry` since it's no longer just the property map portion). It would look somewhat like this:

```rs
struct TargetEntry {
    operation: StandardGate,
    params: SmallVec<[Param; 3]>,
    property_map: IndexMap<Qargs, Option<InstructionProperties>>,
}
```
This structure will later get consumed once it's added to the `Target`. 

We'd keep the API mostly the same, except for the constructor with your suggestions being applicable and change the naming convention from `qk_property_map_{...}` to `qk_target_entry_{...}`. Does this sound good?

Now, since the struct is meant to be consumed by Rust but can only be accessed via pointer, how do we make sure we're using things efficiently? We currently have to clone every mapping we create. Is this something we should be concerned of or is it not as bad as it seems?

Cryoris commented:
That sounds good to me, calling it `TargetEntry` also makes very clear where this is used 👍🏻 

> Now, since the struct is meant to be consumed by Rust but can only be accessed via pointer, how do we make sure we're using things efficiently? We currently have to clone every mapping we create. Is this something we should be concerned of or is it not as bad as it seems?

We have the same issue right now with the `PropertyMap` right now, right? I probably wouldn't worry about this for now, but we could maybe expose a function that consumes the target entry (i.e. frees the pointer).

raynelfss committed:
Refactor: `QkPropsMap` is now `QkTargetEntry`.
- Add rust struct `TargetEntry` to better represent adding a gate to the Target from C,
- Add `SmallVec` as a dependency.
- Remove `qk_target_add_instruction_fixed` in favor of performing the checks during the creation of the `QkTargetEntry`.
- Update documentation with newer additions.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-api

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-api

Cryoris commented on docs/cdoc/qk-target-entry.rst:
```diff
@@ -18,23 +18,20 @@ Here's an example of how this structure works:
-    // Create a Property Map
-    QkPropsMap *property_map = qk_property_map_new();
+    // Create a Property Map for a CX Gate
+    QkTargetEntry *property_map = qk_target_entry_new(QkGate_CX);
```
The comments and variable names need some updating 🙂

raynelfss replied:
Indeed they do :)

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-api

raynelfss committed:
Docs: update documentation to clearly

Cryoris commented on docs/cdoc/qk-target-entry.rst:
```diff
@@ -0,0 +1,37 @@
+.. code-block:: C
+
+    #include <qiskit.h>
+    #include "math.h"
```
```suggestion
    #include <math.h>
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// non-null pointer to a ``QkTargetEntry`` object.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_entry_len(entry: *const TargetEntry) -> usize {
```
Also here, could we call it something that reflects the property better, like `num_instructions` or something? 🙂

raynelfss replied:
For an entry, I think the current name is valid since it is the number of both qargs and properties (we could get away with calling it `qk_target_entry_num_props` or something among those lines). This name, however, could work for the `qk_target_len` method.

raynelfss replied:
Addressed in a35f1bb

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///
+/// @param entry The pointer to the mapping object.
+///
+/// @return The length of the TargetEntry.
```
```suggestion
/// @return The length of the ``QkTargetEntry``.
```

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into target-c-api

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

raynelfss committed:
Chore: Rename len methods.
- Rename `qk_target_len` to `qk_target_num_instructions`.
- Rename `qk_target_entry_len` to `qk_target_num_properties`.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+
+use crate::exit_codes::ExitCode;
+use crate::pointers::{const_ptr_as_ref, mut_ptr_as_ref};
+use core::f64;
```
Just a question: is this the same as `std::f64`? I'm surprised we need an import to use the native type 🤔

mtreinish replied:
We shouldn't need to import this from `core`. We're not in a no-std build for rust we very explicitly rely on the stdlib in cext (and qiskit more broadly). You should just be able to use the built in f64 intrinsic. If that isn't sufficient you might want to use `std::ffi::c_double` but I don't think think that should be necessary.

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+use smallvec::{smallvec, SmallVec};
+
+/// @ingroup QkTarget
+/// Construct a new ``Target`` with the given number of qubits.
```
There are a lot of other other places where this also need updating 
```suggestion
/// Construct a new ``QkTarget`` with the given number of qubits.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_target_new(num_qubits: usize) -> *mut Target {
```
It's a bit strange that the target uses `usize` whereas we always use `u32` as qubit index and as number of qubits in the circuit. Could we make this consistent in the C interface?

raynelfss replied:
We can make it consistent in the Rust API too, this has been a pain point @mtreinish has been bringing up. Fixing...

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// # Example
+///     
+///     QkTarget *target = qk_target_new(5);
+///     uint32_t num_qubits = qk_target_num_qubits(target);
```
Proving my point above 😛

raynelfss replied:
Point taken 😛

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///
+/// @param target A pointer to the ``Target``.
+///
+/// @return The dt value of this ``Target`` or `NaN` if not assigned.
```
```suggestion
/// @return The dt value of this ``Target`` or ``NaN`` if not assigned.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+}
+
+/// @ingroup QkTarget
+/// Returns the ``min_length`` value of this ``Target``.
```
I feel like we should improve the documentation of the target a bit (can be done as follow-up) to explain what granularity, min_length, dt, etc. mean. Right now the C docs don't cover this

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// Sets the dt value of this ``Target``.
+///
+/// @param target A pointer to the ``Target``.
+/// @param dt The ``dt`` value for the system time resolution of input.
```
dt and ``dt`` are used inconsistently

raynelfss replied:
"dt" is the parameter here, while "``dt``" is how it is displayed as the description. Is that not how it's supposed to be used here?

Cryoris replied:
No _that_ is correct, I rather meant in the docs throughout you typically use "dt" but here you used "`dt`" - it's just a minor nit we don't have to change this now either

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_set_dt(target: *mut Target, dt: f64) -> ExitCode {
```
What happens if I would pass `NAN` as value? Do we want to document this value cannot be passed (maybe my preferred solution), or handle it by setting the value to `None`?

raynelfss replied:
Oh, nice catch! I definitely meant to make it be `None` when `NaN` is passed. And I believe rust would handle the `NaN` case here since `f64` has `f64::NaN`

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// Creates an entry to the ``QkTarget`` based on a ``QkGate`` instance with
+/// no parameters.
+///
+/// @param operation The Standard Gate representing this entry in the target.
```
How about this? I'm also fine keeping the original text but with ``QkGate``
```suggestion
/// @param operation The ``QkGate`` whose properties this target entry defines.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///
+/// @param operation The Standard Gate representing this entry in the target.
+///
+/// @return The ``QkTargetEntry`` structure.
```
```suggestion
/// @return A pointer to the new ``QkTargetEntry``.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+
+/// @ingroup QkTargetEntry
+/// Creates an entry to the ``QkTarget`` based on a ``QkGate`` instance with
+/// no parameters.
```
Maybe worth to mention ``qk_target_entry_fixed`` here 🙂

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+}
+
+/// @ingroup QkTargetEntry
+/// Creates an object that will serve as a mapping between an instruction's
```
I feel like these docs should be the same as above, just with the addition that you have fixed parameters -- right now the texts diverge 🙂

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// qargs and instruction properties.
+///
+/// @param operation The Standard Gate representing this entry in the target.
+/// @param params The list of fixed parameters for this gate.
```
This sounds like it would be a list of parameters it is calibrated for, maybe something like this? Might also be good to note the length here, even though it is repeated in the safety section below
```suggestion
/// @param params A pointer to the parameters that the instruction is calibrated for. 
///     The number of parameters must be equal to ``qk_gate_num_params(operation)``.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// matches the the expectations of the standard gate. If the array is insufficently long the
+/// behavior of this function is undefined as this will read outside the bounds of the array.
+/// It can be a null pointer if there are no params for a given gate. You can check
+/// `qk_gate_num_params` to determine how many qubits are required for a given gate.
```
```suggestion
/// ``qk_gate_num_params`` to determine how many qubits are required for a given gate.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+}
+
+/// @ingroup QkTargetEntry
+/// Retrieves the length of the current target entry's property map..
```
```suggestion
/// Retrieves the number of properties stored in the target entry.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///
+/// @param entry The pointer to the mapping object.
+///
+/// @return The length of the ``QkTargetEntry``.
```
```suggestion
/// @return The number of properties in the ``QkTargetEntry``.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// @ingroup QkTargetEntry
+/// Frees the entry.
+///
+/// @note An entry pointer will be freed when added to a ``QkTarget`` via
```
Is it? It doesn't seem like it would be freed, but maybe I'm missing something 🤔 I don't think we have to free upon consumption, maybe we can just update the comment here

raynelfss replied:
In the function ``qk_target_add_instruction`` we take ownership of the pointer to the ``QkTargetEntry`` instance, consuming it. This is what we usually do for the when freeing the pointers.

```rs
#[no_mangle]
#[cfg(feature = "cbinding")]
pub unsafe extern "C" fn qk_target_add_instruction(
    target: *mut Target,
    target_entry: *mut TargetEntry,
) -> ExitCode {
    // SAFETY: Per documentation, the pointer is non-null and aligned.
    let entry = unsafe { Box::from_raw(target_entry) };
    let instruction = entry.operation;
    ...
}
```

Though, now that I look we don't really check the pointer before doing so :)

Cryoris replied:
Oh true, I must've misread! All good then! But yeah it would be good to check the pointer is not null and is aligned 👍🏻

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///
+/// @note Adding parametric gates with non-fixed parameters is currently
+/// not supported. See ``qk_target_add_fixed_instruction()`` for
+/// adding gates with fixed parameters.
```
This is no longer needed

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// @param target A pointer to the ``Target``.
+/// @param target_entry A pointer to the ``TargetEntry`` containing the gate and
+/// the mapping of qargs and InstructionProperties to be associated with this
+/// instruction. The pointer gets freed when added to the ``Target``.
```
This needs some updating, "InstructionProperties" is not an object the user knows about 🙂

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///     QkTargetEntry *entry = qk_target_entry_new(QkGate_CX);
+///     uint32_t qargs[2] = {0, 1};
+///     qk_target_entry_add_property(entry, qargs, 2, 0.0, 0.1);
+///     qk_target_add_instruction(target, entry);
```
It would be nice to assign the outcome to an exit code 🙂

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// @param instruction The instruction to modify.
+/// @param qargs The pointer to the array of ``uint32_t`` values to use as
+/// qargs. Can be ``NULL`` if global.
+/// @param num_qubits The number of qubits in the array.
```
```suggestion
/// @param num_qubits The number of qubits of the instruction.
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+///
+/// The ``qargs`` type is expected to be a pointer to an array of ``u32int_t`` where the length
```
```suggestion
/// The ``qargs`` type is expected to be a pointer to an array of ``uint32_t`` where the length
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+/// matches is specified by ``num_qubits`` and has to match the expectation of the gate. If the
+/// array is insufficently long the behavior of this function is undefined as this will read
+/// outside the bounds of the array. It can be a null pointer if there are no qubits for
+/// a given gate. You can check `qk_gate_num_qubits` to determine how many qubits are required
```
```suggestion
/// a given gate. You can check ``qk_gate_num_qubits`` to determine how many qubits are required
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+}
+
+/// @ingroup QkTarget
+/// Get the length of the ``Target``.
```
Could you rewrite this for number of instructions?

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+///     QkTargetEntry *target_enty = qk_target_entry_new(QkGate_H);
+///     qk_target_add_instruction(target, target_entry);
+///
+///     qk_target_num_instructions(target)
```
```suggestion
///     size_t num_instructions = qk_target_num_instructions(target);
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+    target.len()
+}
+
+// Helpers
```
Personally I'm not a big fan of sections in the code since the code might get shuffled, but you can ofc keep it if you think it is helpful 🙂 
```suggestion
```

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -0,0 +1,777 @@
+    if qargs.is_null() {
+        Qargs::Global
+    } else {
+        // SAFETY: Per the documentation the qubits pointer is non-null and points to an array of num_qubits elements
```
Not sure what the Rust norms are here, but it couldn't hurt to put this statement as `# Safety` section in the docstring 🙂

raynelfss replied:
In here it doesn't matter too much since these are helper functions, but I'll add it to make things better.

Cryoris commented on docs/cdoc/qk-target-entry.rst:
```diff
@@ -0,0 +1,37 @@
+    // Create a Target Entry for a CX Gate
+    QkTargetEntry *entry = qk_target_entry_new(QkGate_CX);
+
+    // Add mapping between (0,1) and InstructionProperties(10e-9, NAN)
```
InstructionProperties is not exposed 🙂

Cryoris replied:
It would be nice to expand a bit here, but this can be left as follow up (like the other comment on docs)

Cryoris commented on docs/cdoc/qk-target-entry.rst:
```diff
@@ -0,0 +1,37 @@
+
+    typedef struct QkTargetEntry QkTargetEntry
+
+A mapping of qargs and instruction properties representing gate map of the 
```
I don't think "qargs" is an official word, maybe we could just use "qubits"?

Cryoris commented on docs/cdoc/qk-target.rst:
```diff
@@ -0,0 +1,50 @@
+A mapping of instructions and properties representing the partiucular constraints
+of a backend. Its purpose is to provide the compiler with information that allows it
+to compile an input circuit into another that is optimized taking in consideration the
+``Target``'s specifications. This structure represents a low level interface to the main
```
```suggestion
``QkTarget``'s specifications. This structure represents a low level interface to the main
```

Cryoris commented on docs/cdoc/qk-target.rst:
```diff
@@ -0,0 +1,50 @@
+    qk_target_add_instruction(target, entry);
+
+    // Add a global H gate
+    qk_target_add_instruction(target, qk_target_entry_new(QkGate_H));
```
I think we have to free the entry (but relates to another question I had above)

raynelfss replied:
The entry gets freed once it gets sent to the `Target`. See [this comment](https://github.com/Qiskit/qiskit/pull/14248#discussion_r2124501314).

Cryoris commented on docs/cdoc/qk-target.rst:
```diff
@@ -0,0 +1,50 @@
+
+The Target C API currently only supports additions of ``StandardGate`` instances as
+``NormalOperation``. The functionality will be expanded over time as we improve our
+Rust data model capabilities.
```
This needs some updating

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,303 @@
+}
+
+/**
+ * Test construction of PropsMap
```
```suggestion
 * Test construction of QkTargetEntry
```

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,303 @@
+ * Test updating an instruction property in the Target using
+ * `update_instruction_property`.
+ */
+int test_target_update_instruction(void) {
```
This is already nice but it would be good to also check the other error codes are raised as expected 🙂

raynelfss replied:
I believe I covered all the possible errors here.

Cryoris replied:
Hm I don't see `QkExitCode_TargetQargMismatch` or `QkExitCode_TargetInstAlreadyExists` being used anywhere, but maybe I'm missing something 🤔

raynelfss replied:
When I mention all possible errors, what I meant is all possible errors thrown by this function, which should be 303 and 304.

Cryoris replied:
Ok sorry my initial comment was not clear enough: I meant it would be good to have tests for all types of ExitCodes we have defined 🙂

Cryoris reviewed (commented):
I'm very happy with the code, but the docs need some more attention (and I think we can complete the reno now 🙂). I also left a comment on the tests 🙂

raynelfss committed:
Chore: Fix documentation further
- Use `u32` for `Target` length and `min_length`.

Co-authored-by: Julien Gacon <gaconju@gmail.com>

raynelfss committed:
Docs: Add example to release note

Cryoris commented on crates/cext/src/exit_codes.rs:
```diff
@@ -43,6 +44,20 @@ pub enum ExitCode {
+    /// An attempt to add a parametric gate without fixed parameters.
+    TargetNonFixedParametricGate = 305,
+    /// An attempt to add a non-parametric gate with parametric methods
+    TargetNonParametricError = 306,
```
Are these 2 actually used anywhere?

raynelfss replied:
I think not anymore, I will remove them.

raynelfss replied:
Removed in 661c8c5

mtreinish commented:
https://github.com/Qiskit/qiskit/pull/14526 should fix the CI failures

raynelfss committed:
Merge branch 'main' into target-c-api

raynelfss committed:
Chore: Remove unused error codes

raynelfss committed:
Fix: Remove conversion in vf2 vf2_layout
- Fix indentation in helper function docstrings in target.rs

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris commented on docs/cdoc/qk-target-entry.rst:
```diff
@@ -21,11 +21,11 @@ Here's an example of how this structure works:
     QkTargetEntry *entry = qk_target_entry_new(QkGate_CX);
 
-    // Add mapping between (0,1) and InstructionProperties(10e-9, NAN)
+    // Add mapping between (0,1) and properties duration of 10e-9 and no error.
```
```suggestion
    // Add mapping between (0,1) and properties duration of 10e-9 and unknown error.
```

raynelfss committed:
Fix: Make `qk_target_entry_add_property` fallible.
- `qk_target_entry_add_property` will throw an error code if the wrong qargs number is ever provided.
- Add comparison of error codes in unittests.

raynelfss committed:
Update docs/cdoc/qk-target-entry.rst

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,357 @@
+ * Test empty constructor for Target
+ */
+int test_empty_target(void) {
+    QkTarget *target = qk_target_new(0);
```
This target is not freed at the end of the test

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,357 @@
+    const uint32_t p_alignment = 4;
+    const uint32_t a_alignment = 1;
+
+    QkTarget *target = qk_target_new(num_qubits);
```
... or this 🙂

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,357 @@
+    if (result_cx_props != QkExitCode_Success) {
+        printf("Unexpected error occurred when adding property to a CX gate entry.");
+        return EqualityError;
+        goto cleanup;
```
This doesn't look quite right (below as well), did you mean to set the result to `EqualityError`? Also we need to free `cx_entry` if we terminate early

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,357 @@
+        goto cleanup;
+    }
+    current_size = qk_target_num_instructions(target);
+    if (current_num_qubits != 2) {
```
Should this check `current_size`?

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,357 @@
+        return EqualityError;
+        goto cleanup;
+    }
+    // CX Gate is not paramtric.
```
Dangling comment

Cryoris commented on test/c/test_target.c:
```diff
@@ -0,0 +1,357 @@
+        goto cleanup;
+    }
+    current_size = qk_target_num_instructions(target);
+    if (current_num_qubits != 3) {
```
`current_size`?

raynelfss committed:
Fix: Check pointer before consuming in `qk_target_add_instruction`.

raynelfss committed:
Chore: Clean up the  target tests.

raynelfss committed:
Fix: More mistakes in tests.

Cryoris reviewed (approved):
LGTM thanks for all the work -- I believe we're in very good shape now and have a API that works well for C space. The true test will come once users will want to build custom targets (assuming that auto-generated targets from backends will be generated in Rust) 🙂
